File: SpringBootCache/src/main/java/com/example/component/SimpleBookRepository.java
package com.example.component;

import com.example.domain.Book;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Component;

@Component
public class SimpleBookRepository implements BookRepository {

	private static final Logger logger = LoggerFactory.getLogger(SimpleBookRepository.class);
	
	@Override
	@Cacheable(value="book", key="#isbn")
	public Book getByIsbn(String isbn) {
		simulateSlowService();
		return new Book(isbn, "Some book");
	}

	private void simulateSlowService() {
		try {
			long time = 3000L;
			Thread.sleep(time);
		} catch (InterruptedException e) {
			throw new IllegalStateException(e);
		}
	}

	@Override
	@CacheEvict(value="book", key="#isbn")
	public void refresh(String isbn) {
		logger.info("cache clear => " + isbn);
	}
}

File: SpringBootCache/src/main/kotlin/com/example/kotlin/component/SimpleBookRepository.kt
package com.example.kotlin.component

import com.example.kotlin.domain.Book
import org.slf4j.LoggerFactory
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Component

@Component
class SimpleBookRepository : BookRepository {

    companion object {
        private val logger = LoggerFactory.getLogger(SimpleBookRepository::class.java)
        private const val CACHE_BOOK = "book"
    }

    @Cacheable(value = [CACHE_BOOK], key = "#isbn")
    override fun getByIsbn(isbn: String): Book {
        simulateSlowService()
        return Book(isbn, "Some book")
    }

    private fun simulateSlowService() {
        try {
            Thread.sleep(3000L)
        } catch (e: InterruptedException) {
            throw IllegalStateException(e)
        }

    }

    @CacheEvict(value = [CACHE_BOOK], key = "#isbn")
    override fun refresh(isbn: String) {
        logger.info("cache clear => $isbn")
    }
}

